### DISCUSSION TITLE
Caching behavior across different pages

### DISCUSSION THREAD

**Discussion by hosseint79** [TanStack/query#8141] [Q&A]
### Describe the bug

I am using Next.js with React Query and have encountered an issue with query caching behavior across different pages.

### Your minimal, reproducible example

haven't

### Steps to reproduce

I have two pages: page1 and page2.

Both pages use useQuery with the same query key.

The useQuery configuration for both pages has staleTime set to 0 and cacheTime (or gcTime) set to 0.

When I navigate from page1 to page2, the following unexpected behaviors occur:

The data queried on page1 does not seem to be garbage collected, even though cacheTime is 0.

On page2, instead of creating a new instance of the query, it appears to reuse the previous data.

### Expected behavior

When navigating away from page1, the query data should be immediately removed from the cache due to cacheTime set to 0.
When navigating to page2, useQuery should initiate a fresh fetch since the data should have been garbage collected.

### How often does this bug happen?

Every time

### Platform

os: mac
browser: chrome

### Tanstack Query adapter

None

### TanStack Query version

5.40.0

**Comment by TkDodo** [comment]
sounds related:
- https://github.com/TanStack/query/issues/8136

but hard to say without seeing a reproduction

**Reply by hosseint79** [reply]
Thank you for your response! I’ve created a sandbox demo and recorded a video demonstrating the issue in more detail.

https://github.com/user-attachments/assets/0241667f-dc0e-48d6-941f-17294b56be25

Demo:
https://codesandbox.io/p/devbox/nervous-torvalds-twgwjt

I’d like to better understand why this behavior occurs. Specifically, why does the data persist across pages when both cacheTime and staleTime are set to 0? It seems like the query should be garbage collected immediately when leaving page1, and page2 should create a new instance. Could you explain how React Query is handling this scenario?

**Reply by TkDodo** [reply]
sandbox is private so I can’t look at it, but anyways, if you switch between pages and the next page immediately uses the same query, I think there is no time for us to garbage collect, so data stays in the cache.

Just to be sure: In v5, the setting is `gcTime`, not `cacheTime` (not sure how you have it configured because I can’t see the sandbox and your video doesn’t show it...
